### PR TITLE
optimize vbft submit block

### PR DESCRIPTION
**Pull request by xiemylogos**

### CODE REVIEW

File: consensus/vbft/service.go [hunk 1]
```diff
@@ -1043,6 +1055,32 @@ func (self *Server) onConsensusMsg(peerIdx uint32, msg ConsensusMsg, msgHash com
...
+		}
+		submitNum := self.msgPool.GetBlockSubmitMsgNums(msgBlkNum)
+		m := len(self.config.Peers) - (len(self.config.Peers)*3)/7
+		if submitNum < uint32(m) {
+			return
+		}
+		log.Infof("submit msg  submitNum:%d,m:%d", submitNum, m)
+		if err := self.chainStore.SubmitBlock(msgBlkNum); err != nil {
+			log.Errorf("SubmitBlock err:%s", err)
+		}
```
Comment (laizy) [inline consensus/vbft/service.go:1093]: where is the state root comparison between remote's and local's?

File: consensus/vbft/chain_store.go [hunk 2]
```diff
@@ -154,6 +142,23 @@ func (self *ChainStore) AddBlock(block *Block) error {
 	return nil
 }
 
+func (self *ChainStore) SubmitBlock(blkNum uint32) error {
+	if blkNum == 0 {
+		return nil
+	}
+	if submitBlk, present := self.pendingBlocks[blkNum]; submitBlk != nil && submitBlk.needSubmitBlock == false && present {
```
Comment (laizy) [inline consensus/vbft/chain_store.go]: why `needSubmitBlock == false`?

Comment (xiemylogos) [reply consensus/vbft/chain_store.go]: yes,I will modify param name in next commit. needSubmitedBlock

Comment (laizy) [reply consensus/vbft/chain_store.go]: rename to `hasSubmitted`

Comment (xiemylogos) [reply consensus/vbft/chain_store.go]: ok

File: consensus/vbft/service.go (hunk 1 above)
Comment (xiemylogos) [reply consensus/vbft/service.go:1093]: already added  in next commit

File: consensus/vbft/service.go [hunk 3]
```diff
@@ -229,6 +229,25 @@ func (self *Server) handleBlockPersistCompleted(block *types.Block) {
...
+func (self *Server) SubmitBlock(blkNum uint32, stateRoot common.Uint256) {
+	cMsgs := self.msgPool.GetBlockSubmitMsgNums(blkNum)
+	m := len(self.config.Peers) - (len(self.config.Peers)*3)/7
+	if len(cMsgs) < m {
+		return
+	}
+	for _, msg := range cMsgs {
+		c := msg.(*blockSubmitMsg)
+		if c != nil {
+			if c.BlockStateRoot != stateRoot {
```
Comment (laizy) [inline consensus/vbft/service.go]: why not check it before `AddMsg`?

Comment (xiemylogos) [reply consensus/vbft/service.go]: there are some concurrency problem when  call GetExecMerkleRoot() function,so get merkroot in same goroutine.

File: consensus/vbft/service.go [hunk 4]
```diff
@@ -1337,6 +1377,14 @@ func (self *Server) processMsgEvent() error {
...
 				// FIXME: add msg from msg-pool to block-pool when starting new block-round
 			}
+		case BlockSubmitMessage:
+			pMsg := msg.(*blockSubmitMsg)
+			msgBlkNum := pMsg.GetBlockNum()
+			stateRoot, err := self.chainStore.GetExecMerkleRoot(msgBlkNum)
+			if err != nil {
+				return nil
+			}
+			self.SubmitBlock(msgBlkNum, stateRoot)
```
Comment (Honglei-Cong) [inline consensus/vbft/service.go]: do block submission in actionLoop?

File: consensus/vbft/block_pool.go [hunk 5]
```diff
@@ -664,6 +664,15 @@ func (pool *BlockPool) setBlockSealed(block *Block, forEmpty bool, sigdata bool)
...
 
+	stateRoot, err := pool.chainStore.GetExecMerkleRoot(c.SealedBlock.getBlockNum() - 1)
+	if err != nil {
+		log.Errorf("handleBlockSubmit failed:%s", err)
+		return nil
+	}
+	if blocksubmitMsg, _ := pool.server.constructBlockSubmitMsg(c.SealedBlock.getBlockNum()-1, stateRoot); blocksubmitMsg != nil {
+		pool.server.broadcast(blocksubmitMsg)
+	}
+	pool.server.SubmitBlock(c.SealedBlock.getBlockNum()-1, stateRoot)
```
Comment (Honglei-Cong) [inline consensus/vbft/block_pool.go]: change to send a request to actionLoop?

Comment (xiemylogos) [reply consensus/vbft/block_pool.go]: ok in next commit

File: consensus/vbft/service.go (hunk 4 above)
Comment (xiemylogos) [reply consensus/vbft/service.go]: ok in next commit

File: consensus/vbft/service.go [hunk 6]
```diff
@@ -1043,6 +1064,30 @@ func (self *Server) onConsensusMsg(peerIdx uint32, msg ConsensusMsg, msgHash com
...
 			msg:      msg,
 		}
+	case BlockSubmitMessage:
+		pMsg, ok := msg.(*blockSubmitMsg)
+		if !ok {
+			log.Error("invalid msg with submit msg type")
+			return
+		}
+		msgBlkNum := pMsg.GetBlockNum()
+		if msgBlkNum > self.GetCurrentBlockNo() {
```
Comment (Honglei-Cong) [inline consensus/vbft/service.go:1074]: ignore the msg if node has  submitted the block?

**Review by laizy — DISMISSED** [review]

File: consensus/vbft/service.go (hunk 6 above)
Comment (xiemylogos) [reply consensus/vbft/service.go:1074]: already added,in new commit

**Review by Honglei-Cong — APPROVED** [review]

**Pull request merged**